File: NetDeck/StrengthStats.m
//
//  StrengthStats.m
//  Net Deck
//
//  Created by Gereon Steffens on 15.02.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "StrengthStats.h"

@implementation StrengthStats

-(StrengthStats*) initWithDeck:(Deck *)deck
{
    if ((self = [super init]))
    {
        // calculate strength distribution
        NSMutableDictionary* strengths = [NSMutableDictionary dictionary];
        for (CardCounter* cc in deck.cards)
        {
            NSInteger str = cc.card.strength;
            if (str != -1)
            {
                NSNumber* n = [strengths objectForKey:@(str)];
                int prev = n == nil ? 0 : [n intValue];
                n = @(prev + cc.count);
                [strengths setObject:n forKey:@(str)];
            }
        }
        
        NSArray* sections = [[strengths allKeys] sortedArrayUsingComparator:^(NSNumber* n1, NSNumber* n2) { return [n1 compare:n2]; }];
        NSMutableArray* values = [NSMutableArray array];
        for (NSNumber*n in sections)
        {
            [values addObject:[strengths objectForKey:n]];
        }
        NSAssert(sections.count == values.count, @"");
        self.tableData = [[TableData alloc] initWithSections:sections andValues:values];
    }
    return self;
}

-(CGFloat) height
{
    return self.tableData.sections.count == 0 ? 0 : 300;
}

-(CPTGraphHostingView*) hostingView
{
    return [self hostingViewForDelegate:self identifier:@"Strength"];
}

#pragma mark - CPTPlotDataSource methods

-(CPTLayer *)dataLabelForPlot:(CPTPlot *)plot recordIndex:(NSUInteger)index
{
    static CPTMutableTextStyle *labelText = nil;
    
    if (!labelText)
    {
        labelText = [[CPTMutableTextStyle alloc] init];
        labelText.color = [CPTColor blackColor];
    }
    
    NSNumber* strength = [self.tableData.sections objectAtIndex:index];
    NSNumber* cards = [self.tableData.values objectAtIndex:index];
    
    NSString* str = nil;
    int c = [cards intValue];
    if (c > 0)
    {
        str = [NSString stringWithFormat:l10n(@"Strength %d\n%d %@"), [strength intValue], c, CARDS(c)];
    }
    
    // 5 - Create and return layer with label text
    return [[CPTTextLayer alloc] initWithText:str style:labelText];
}

@end

File: NetDeck/StrengthStats.swift
//
//  StrengthStats.swift
//  NetDeck
//
//  Created by Gereon Steffens on 25.12.16.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import CorePlot

class StrengthStats: Stats {
    
    init(deck: Deck) {
        super.init()
        self.useBlues = true
        
        var strengths = [String: Int]()
        for cc in deck.cards.filter({ $0.card.strength >= 0 }) {
            let str = "\(cc.card.strength)"
            strengths[str] = (strengths[str] ?? 0) + cc.count
        }
        
        let sections = strengths.keys.sorted { $0 < $1 }
        let values = sections.map { [strengths[$0]!] }
        
        assert(sections.count == values.count)
        
        self.tableData = TableData(sections: sections, values: values)
    }
    
    var hostingView: CPTGraphHostingView {
        return self.hostingView(for: self, identifier: "Strength")
    }
    
    func dataLabel(for plot: CPTPlot, record index: UInt) -> CPTLayer? {
        let strength = self.tableData.sections[Int(index)]
        let cards = self.tableData.values[Int(index)][0]
    
        let str = String(format: "Strength %@\n%d %@".localized(), strength, cards, cardsString(cards))
        let style = CPTMutableTextStyle()
        style.color = CPTColor(uiColor: .label)
        return CPTTextLayer(text: str, style: style)
    }
}
